File: 卡码/代码随想录/1/10-单调栈/Go/LeetCode_0042.go
// 42. 接雨水
package main

import (
	"math"
)

func trap(height []int) int {
	res := 0
	st := make([]int, 0)
	for index, num := range height {
		for len(st) > 0 && num > height[st[len(st)-1]] {
			low := st[len(st)-1]
			st = st[:len(st)-1]
			if len(st) > 0 {
				h := int(math.Min(float64(num), float64(height[st[len(st)-1]]))) - height[low]
				w := index - st[len(st)-1] - 1
				res += h * w
			}
		}
		st = append(st, index)
	}
	return res
}

File: 卡码/代码随想录/1/10-单调栈/Js/LeetCode_0042.js
//42. 接雨水
const trap = function(height) {
	const st=new Array();
	if(height.length<3) return 0;
	let res=0;
	for(let i=0;i<height.length;i++){
		while(st.length&&height[i]>height[st[st.length-1]]){
			let low=st.pop();
			if(st.length){
				let h=Math.min(height[i],height[st[st.length-1]])-height[low];
				let w=i-st[st.length-1]-1;
				res+=h*w;
			}
		}
		st.push(i);
	}
	return res;
};
